Source/Core/Core/PowerPC/Jit64/Jit64_Util.cpp

```diff
@@ -0,0 +1,420 @@
+// Copyright 2009 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Core/HW/GPFifo.h"
+#include "Core/PowerPC/Jit64/Jit.h"
+
+using namespace Gen;
+
+// This set of functions encapsulates the entire craziness we need to deal
+// with when dealing with loads and stores. When running in MMU mode, we need
+// to ensure that all modifications made to input registers (other than the
+// write to a register as part of a load) are rolled back in the case of
+// exception. In some cases we use a scratch register for this, in other cases
+// we create a register transaction that will restore the register value.
+
+void Jit64::SafeWrite(GPRRegister& reg_value, GPRRegister& reg_addr, GPRRegister& offset,
+                      int accessSize, bool swap, bool update)
+{
+  _assert_msg_(DYNAREC, accessSize == 8 || accessSize == 16 || accessSize == 32 || accessSize == 64,
+               "Invalid write size: %d", accessSize);
+
+  if (update && !reg_addr.HasPPCRegister())
+  {
+    // This form of instruction is illegal
+    WriteInvalidInstruction();
```

Comment:
I should make this a debug assert - it is more of a developer sanity check that the caller is passing something that makes sense. It doesn't make sense to do this in the callers, since most of them either pass a value unconditionally or do something like `single ? 32 : 64`.